piptools/resolver.py

```diff
@@ -219,14 +219,16 @@ def _group_constraints(self, constraints):
             flask~=0.7
 
         """
-        for _, ireqs in full_groupby(constraints, key=key_from_ireq):
+        for _, ireqs in groupby(
+            sorted(
+                constraints,
+                # Within each group, sort by editables first, and within that,
+                # non-constraints before constraints (False sorts before True):
+                key=lambda x: (key_from_ireq(x), not x.editable, x.constraint),
+            ),
+            key=key_from_ireq,
+        ):
```

Comment:
AFAIK `sort` and `groupby` by different key functions could be a potential bug.

Reply:
> The operation of groupby() is similar to the uniq filter in Unix. It generates a break or new group every time the value of the key function changes (which is why it is usually necessary to have sorted the data using the same key function). That behavior differs from SQL’s GROUP BY which aggregates common elements regardless of their input order.

https://docs.python.org/3/library/itertools.html#itertools.groupby

Reply:
I believe in this case, because the sort key is the group key plus extra tie-breakers, it should satisfy the needs of the `groupby`. The `groupby` cares about comparing the group key between elements. In this way, we can sort within the groups when we sort everything. Alternatively, we could sort each group again after the sort+groupby. 

Ultimately, we should decide if this sort order is even necessary - I added it only for backwards compatibility with the comment I removed about yielding just the editable. Now it yields a copy of the editable with updates from others in the group. But we could implement that however, if it's desirable.

Reply:
In fact, we may don't touch these lines and everything would work. The **order** of the sort doesn't matter actually, so these changes can be reverted.

Reply:
The point is that the fact of *sort* is matters, not order.